[FLINK-30418] Implement synchronous KinesisClient in EFO

darenwkt opened the pull request:
Implemented synchronous KinesisClient as described in https://issues.apache.org/jira/browse/FLINK-30418.



KinesisProxyV2 supports the following operations:

    subscribeToShard
    describeStreamSummary
    describeStreamConsumer
    registerStreamConsumer
    deregisterStreamConsumer

 

All the above operations are currently performed via an AsyncHTTPClient. However, only subsribeToShard needs to be asynchronous, the rest are called synchronously. This PR updates all the operation to use a SyncHTTPClient except subscribeToShard.

Change Summary:

    subscribeToShard: async (no change)
    describeStreamSummary: async -> sync
    describeStreamConsumer: async -> sync
    registerStreamConsumer: async -> sync
    deregisterStreamConsumer: async -> sync


Tested in local environment with a sample input and output Kinesis Stream on local flink cluster.

dannycranmer commented:
@darenwkt did you verify this works using EFO? There are no e2e tests for EFO since Kinesalite does not support it

dannycranmer commented on flink-connector-aws-base/pom.xml:
```diff
@@ -52,6 +52,11 @@ under the License.
             <artifactId>netty-nio-client</artifactId>
         </dependency>
 
+        <dependency>
+            <groupId>software.amazon.awssdk</groupId>
+            <artifactId>apache-client</artifactId>
+        </dependency>
```
We will need to ensure this is shaded correctly and NOTICE files are updated

darenwkt replied:
Verified that apache-client is in the installed JAR.

<img width="461" alt="Screenshot 2022-12-16 at 19 51 45" src="https://user-images.githubusercontent.com/108749182/208178095-b0490c6f-67c7-4b59-90e2-811415ad2004.png">


Verified that NOTICE files are updated.

```

- software.amazon.awssdk:apache-client:2.17.247
- software.amazon.awssdk:netty-nio-client:2.17.247
```

dannycranmer replied:
NOTICE files do not update automatically, so either:
- It is already included and this is not necessary, or
- The NOTICE files need updating

Since flink-connector-aws-base is common to all connectors we will need to update all the NOTICE files I believe.

dannycranmer replied:
Yes, I already see it. For all connectors, thanks

dannycranmer commented on flink-connector-aws-base/src/main/java/org/apache/flink/connector/aws/util/AWSAsyncSinkUtil.java:
```diff
@@ -161,4 +163,86 @@ S createAwsAsyncClient(
+                    S extends SdkClient,
+                    T extends
+                            AwsSyncClientBuilder<? extends T, S> & AwsClientBuilder<? extends T, S>>
+            S createAwsSyncClient(
```
Why do we need so many overloaded versions of this? Looks like only 1 is used

darenwkt replied:
Done, have cleaned the rest

dannycranmer commented on flink-connector-aws-base/src/main/java/org/apache/flink/connector/aws/util/AWSAsyncSinkUtil.java:
```diff
@@ -161,4 +163,86 @@ S createAwsAsyncClient(
+        return createAwsSyncClient(configProps, clientBuilder, httpClient, overrideConfiguration);
+    }
+
+    @VisibleForTesting
```
Where is this tested such that it requires this scope?

darenwkt replied:
Done, have clean this up

dannycranmer commented on flink-connector-aws-base/src/main/java/org/apache/flink/connector/aws/util/AWSGeneralUtil.java:
```diff
@@ -301,6 +303,58 @@ public static SdkAsyncHttpClient createAsyncHttpClient(
+    public static SdkHttpClient createSyncHttpClient(
+            final Properties configProperties, final ApacheHttpClient.Builder httpClientBuilder) {
+
+        final AttributeMap.Builder clientConfiguration =
+                AttributeMap.builder().put(SdkHttpConfigurationOption.TCP_KEEPALIVE, true);
+
+        Optional.ofNullable(
+                        configProperties.getProperty(
+                                AWSConfigConstants.HTTP_CLIENT_MAX_CONCURRENCY))
+                .map(Integer::parseInt)
+                .ifPresent(
+                        integer ->
+                                clientConfiguration.put(
+                                        SdkHttpConfigurationOption.MAX_CONNECTIONS, integer));
+
+        Optional.ofNullable(
+                        configProperties.getProperty(
+                                AWSConfigConstants.HTTP_CLIENT_READ_TIMEOUT_MILLIS))
+                .map(Integer::parseInt)
+                .map(Duration::ofMillis)
+                .ifPresent(
+                        timeout ->
+                                clientConfiguration.put(
+                                        SdkHttpConfigurationOption.READ_TIMEOUT, timeout));
+
+        Optional.ofNullable(configProperties.getProperty(AWSConfigConstants.TRUST_ALL_CERTIFICATES))
+                .map(Boolean::parseBoolean)
+                .ifPresent(
+                        bool ->
+                                clientConfiguration.put(
+                                        SdkHttpConfigurationOption.TRUST_ALL_CERTIFICATES, bool));
+
+        Optional.ofNullable(configProperties.getProperty(AWSConfigConstants.HTTP_PROTOCOL_VERSION))
+                .map(Protocol::valueOf)
+                .ifPresent(
+                        protocol ->
+                                clientConfiguration.put(
+                                        SdkHttpConfigurationOption.PROTOCOL, protocol));
+
```
Seems like this is duplicated for Async/Sync. Can we extract to a common method?

darenwkt replied:
Done, extracted to common method

dannycranmer commented on flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/proxy/KinesisProxyV2Factory.java:
```diff
@@ -55,25 +58,39 @@ public static KinesisProxyV2Interface createKinesisProxyV2(final Properties conf
+                        KinesisStreamsConfigConstants.BASE_KINESIS_USER_AGENT_PREFIX_FORMAT,
+                        KinesisStreamsConfigConstants.KINESIS_CLIENT_USER_AGENT_PREFIX);
+
+        return new KinesisProxyV2(
+                asyncClient, client, asyncHttpClient, httpClient, configuration, BACKOFF);
```
As we discussed, we should split `KinesisProxyV2` so we do not need to create both client. In the case of registerStreamConsumer it will not use the async client and it is expensive to create a thread pool, then trash it. We should only create the async client when we actually need to use it

darenwkt replied:
Thank you for clarifying, I have split `KinesisProxyV2` to `KinesisProxySyncV2` and `KinesisProxyAsyncV2`. 

For StreamRegistrar related operations (i.e registerStreamConsumer), only `KinesisProxySyncV2` will be created. 
For EFO related operations (i.e subsribeToShard), only `KinesisProxyAsyncV2` will be created.

darenwkt commented:
> @darenwkt did you verify this works using EFO? There are no e2e tests for EFO since Kinesalite does not support it

Yes, I verified by running EFOConsumer examples from https://github.com/aws-samples/amazon-kinesis-data-analytics-examples/blob/master/EfoConsumer/pom.xml against my local Flink 1.16.0 cluster.

dannycranmer commented on flink-connector-aws-base/src/main/java/org/apache/flink/connector/aws/util/AWSAsyncSinkUtil.java:
```diff
@@ -161,4 +163,47 @@ S createAwsAsyncClient(
     }
+
+    /**
+     * @param configProps configuration properties
```
This javadoc is missing a comment.

darenwkt replied:
Done added

dannycranmer commented on flink-connector-aws-base/src/main/java/org/apache/flink/connector/aws/util/AWSAsyncSinkUtil.java:
```diff
@@ -161,4 +163,47 @@ S createAwsAsyncClient(
+                    final String awsClientUserAgentPrefix) {
+        SdkClientConfiguration clientConfiguration = SdkClientConfiguration.builder().build();
+
+        String flinkUserAgentPrefix =
+                Optional.ofNullable(configProps.getProperty(awsClientUserAgentPrefix))
+                        .orElse(
+                                formatFlinkUserAgentPrefix(
+                                        awsUserAgentPrefixFormat + V2_USER_AGENT_SUFFIX));
+
+        final ClientOverrideConfiguration overrideConfiguration =
+                createClientOverrideConfiguration(
+                        clientConfiguration,
+                        ClientOverrideConfiguration.builder(),
+                        flinkUserAgentPrefix);
+
+        if (configProps.containsKey(AWSConfigConstants.AWS_ENDPOINT)) {
+            final URI endpointOverride =
+                    URI.create(configProps.getProperty(AWSConfigConstants.AWS_ENDPOINT));
+            clientBuilder.endpointOverride(endpointOverride);
+        }
+
```
This code is duplicated in this file, can you extract to a method please

darenwkt replied:
Done refactored

dannycranmer commented on flink-connector-kinesis/pom.xml:
```diff
@@ -256,6 +256,11 @@ under the License.
             <artifactId>netty-nio-client</artifactId>
         </dependency>
 
+        <dependency>
+            <groupId>software.amazon.awssdk</groupId>
+            <artifactId>apache-client</artifactId>
+        </dependency>
+
```
Why is this needed explicitly in here and `aws-base`?

darenwkt replied:
Got it, removing this and keep it in aws-base

dannycranmer commented on flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/publisher/fanout/StreamConsumerRegistrar.java:
```diff
@@ -77,8 +79,8 @@ public StreamConsumerRegistrar(
     public String registerStreamConsumer(final String stream, final String streamConsumerName)
-            throws ExecutionException, InterruptedException {
-        LOG.debug("Registering stream consumer - {}::{}", stream, streamConsumerName);
+            throws Exception {
```
Why did you make this change? This will possible change the error handling behaviour

darenwkt replied:
The main reason of the change is that we are using SyncClient for registerStreamConsumer, so it is not a FutureTask and it won't throw ExecutionException. However, I agree that InterruptedException should be readded as it might be thrown during registrationBackoff

vahmed-hamdy replied:
I agree with @dannycranmer , we would better not generalize the Exception.

darenwkt replied:
I think we can generalise it here because previously ExecutionException, InterruptedException were thrown from a CompletableFuture get method. However, we are no longer using Future here as registerStreamConsumer is called synchronously now (via Callable call method) instead of async. Unlike CompletableFuture, Callable call method throws Exception instead of Execution + InterruptedException.

dannycranmer replied:
My concerns is the Exception is a wider catch bucket than before. However given `ExecutionException, InterruptedException` would have caught the majority and the parent wraps in a `FlinkKinesisStreamConsumerRegistrarException`, it looks ok

dannycranmer commented on flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/publisher/fanout/StreamConsumerRegistrar.java:
```diff
@@ -77,8 +79,8 @@ public StreamConsumerRegistrar(
-            throws ExecutionException, InterruptedException {
-        LOG.debug("Registering stream consumer - {}::{}", stream, streamConsumerName);
+            throws Exception {
+        LOG.warn("Registering stream consumer - {}::{}", stream, streamConsumerName);
```
Why is this a warning?

darenwkt replied:
Done, fixed this

dannycranmer commented on flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/publisher/fanout/StreamConsumerRegistrar.java:
```diff
@@ -90,7 +92,7 @@ public String registerStreamConsumer(final String stream, final String streamCon
         String streamArn = describeStreamSummaryResponse.streamDescriptionSummary().streamARN();
 
-        LOG.debug("Found stream ARN - {}", streamArn);
+        LOG.warn("Found stream ARN - {}", streamArn);
```
Why is this a warning?

darenwkt replied:
Done, fixed this

dannycranmer commented on flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/publisher/fanout/StreamConsumerRegistrar.java:
```diff
@@ -243,28 +244,28 @@ private void waitForConsumerToDeregister(
-        } catch (ExecutionException ex) {
+            response = responseSupplier.call();
+        } catch (Exception ex) {
+            LOG.warn("describeStreamConsumer caught ExecutionException: {}", ex);
```
`caught ExecutionException` not necessarily now you are catching `Exception`. This change is slightly worrying, what happens now if an `InterruptedException` is thrown?

dannycranmer commented on flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/proxy/KinesisProxySyncV2.java:
```diff
@@ -148,11 +139,10 @@ public RegisterStreamConsumerResponse registerStreamConsumer(
                         .consumerName(consumerName)
                         .build();
 
+        LOG.warn("registerStreamConsumer with arn: {}, consumerName: {}", streamArn, consumerName);
```
Why is this a warning?

dannycranmer commented on flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/proxy/KinesisProxySyncV2.java:
```diff
@@ -114,6 +104,7 @@ public DescribeStreamConsumerResponse describeStreamConsumer(
                         .streamARN(streamArn)
                         .consumerName(consumerName)
                         .build();
+        LOG.warn("describeStreamConsumer with arn: {}, consumerName: {}", streamArn, consumerName);
```
Why is this a warning? (did you leave dev debug logs in?)

darenwkt replied:
Yes, my apology for missing this in post-cleanup, doing a round of check again to remove them

dannycranmer commented on flink-connector-kinesis/src/test/java/org/apache/flink/streaming/connectors/kinesis/proxy/KinesisProxyAsyncV2Test.java:
```diff
@@ -0,0 +1,171 @@
+import static org.apache.flink.streaming.connectors.kinesis.config.ConsumerConfigConstants.SUBSCRIBE_TO_SHARD_BACKOFF_BASE;
+import static org.apache.flink.streaming.connectors.kinesis.config.ConsumerConfigConstants.SUBSCRIBE_TO_SHARD_BACKOFF_EXPONENTIAL_CONSTANT;
+import static org.apache.flink.streaming.connectors.kinesis.config.ConsumerConfigConstants.SUBSCRIBE_TO_SHARD_BACKOFF_MAX;
+import static org.mockito.ArgumentMatchers.eq;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.verify;
```
Mockito is [banned from use](https://flink.apache.org/contributing/code-style-and-quality-common.html#avoid-mockito---use-reusable-test-implementations) in Flink. I know we do already use it in places but we should not use for new tests. Can we refactor to reusable test implementations?

darenwkt replied:
Have refactored this away from mockito

vahmed-hamdy commented on flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/publisher/fanout/StreamConsumerRegistrar.java:
```diff
@@ -218,7 +219,7 @@ private void waitForConsumerToDeregister(
             final String streamConsumerArn,
             final int initialAttempt)
-            throws InterruptedException, ExecutionException {
+            throws Exception {
```
Same goes here

darenwkt replied:
Same reply as above: https://github.com/apache/flink-connector-aws/pull/39/files/de6c5579a3b1d4b59938a7b6c3fcdc56d4ef88d4#r1058982327

vahmed-hamdy commented on flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/proxy/KinesisProxyAsyncV2Interface.java:
```diff
@@ -0,0 +1,41 @@
+            SubscribeToShardRequest request, SubscribeToShardResponseHandler responseHandler);
+
+    /** Destroy any open resources used by the factory. */
+    default void close() {
```
Why do we provide a default no-op method?

darenwkt replied:
In the implementation (i.e KinesisProxyAsyncV2.java), close() is overriden and used to close any open HttpClient gracefully. 

I think it's ok to put as default for cases that do not require closing any resources. But do let me know if you think I should remove the default so that any implementation have to explicitly implement it.

dannycranmer replied:
Seems like a shortcut for testing. I would rather move it to `KinesisProxyAsyncV2InterfaceAdapter` to force developers to think about `close()` in src. We missed this before and it caused issues!

darenwkt replied:
Got it, I have removed default and moved this to KinesisProxyAsyncV2InterfaceAdapter and KinesisProxySyncV2InterfaceAdapter to implement

dannycranmer commented:
@darenwkt please repush with correct component tag in commit message, ie `[FLINK-XXXX][Connectors/Kinesis] ...`

dannycranmer commented on flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/proxy/KinesisProxyAsyncV2.java:
```diff
@@ -0,0 +1,78 @@
+@Internal
+public class KinesisProxyAsyncV2 implements KinesisProxyAsyncV2Interface {
+
+    private static final Logger LOG = LoggerFactory.getLogger(KinesisProxyAsyncV2.class);
```
This is unused

darenwkt replied:
Have removed, thank you

darenwkt commented on flink-connector-aws-base/src/main/java/org/apache/flink/connector/aws/util/AWSAsyncSinkUtil.java:
```diff
@@ -161,4 +160,64 @@ S createAwsAsyncClient(
+                    S extends SdkClient,
+                    T extends
+                            AwsSyncClientBuilder<? extends T, S> & AwsClientBuilder<? extends T, S>>
+            S createAwsSyncClient(
```
I think refactoring it to AWSGeneralUtil will be messy as some Client Creation logic will have to be refactored to AWSGeneralUtil. Can I suggest renaming `AWSAsyncSinkUtil` to `AWSClientUtil` instead? This will be more tidy as all the utils to create client are in this class.

darenwkt commented:
@dannycranmer, @vahmed-hamdy I have addressed the comments and retested locally against a local Flink 1.16 cluster and Kinesis EFO Stream. Appreciate if you could review again, thank you
